libbeat/outputs/tidb_slow_query/client.go

```diff
@@ -0,0 +1,233 @@
+package tidb_slow_query
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/outputs"
+	"github.com/elastic/beats/v7/libbeat/publisher"
+	"github.com/go-sql-driver/mysql"
+	lru "github.com/hashicorp/golang-lru"
+	"time"
+
+	_ "github.com/go-sql-driver/mysql"
+)
+
+const (
+	// 2000 clusters
+	insertStmtCacheSize = 2000
+	noTable             = 1146
+	noPartition         = 1526
+	clusterIDFieldKey   = "kubernetes.namespace"
+	noClusterID         = "NO_CLUSTER_ID"
+)
+
+type client struct {
+	db        *sql.DB
+	conn      *sql.Conn
+	observer  outputs.Observer
+	timeout   time.Duration
+	database  string
+	dsn       string
+	retention int
+	rollStep  int
+	stmtCache *lru.Cache
+	log       *logp.Logger
+}
+
+func newClient(
+	observer outputs.Observer,
+	timeout time.Duration,
+	database, dsn string,
+	retention, rollStep int,
+) (*client, error) {
+	db, err := sql.Open("mysql", dsn)
+	if err != nil {
+		return nil, err
+	}
+	cache, err := lru.New(insertStmtCacheSize)
+	if err != nil {
+		return nil, err
+	}
+	c := &client{
+		observer:  observer,
+		timeout:   timeout,
+		database:  database,
+		db:        db,
+		dsn:       dsn,
+		retention: retention,
+		rollStep:  rollStep,
+		stmtCache: cache,
+		log:       logp.NewLogger("tidb_slow_query"),
+	}
+	return c, nil
+}
+
+// Connect try to create a new connection and replace the conn field
+func (c *client) Connect() error {
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+	if c.conn != nil && c.conn.PingContext(ctx) == nil {
+		return nil
+	}
+	conn, err := c.db.Conn(ctx)
+	if err != nil {
+		return err
+	}
+	c.conn = conn
+	return nil
+}
+
+func (c *client) Close() error {
+	return c.conn.Close()
+}
+
+func (c *client) Publish(ct context.Context, batch publisher.Batch) error {
+	ctx, cancel := context.WithTimeout(ct, c.timeout)
+	defer cancel()
+
+	// todo: support multi events from different cluster (with different table name)
+	event := batch.Events()[0]
+
+	table, err := c.getClusterIDAsTableName(event)
+	if err != nil {
+		c.observer.Dropped(1)
+		batch.Cancelled()
+		return err
+	}
+
+	// get driver statement
+	if !c.stmtCache.Contains(table) {
+		sqlString := insertStmt(c.database, table)
```

Comment:
The prepared statement can be cached and reused, which can improve performance.

Reply:
There is a cache at line #107.